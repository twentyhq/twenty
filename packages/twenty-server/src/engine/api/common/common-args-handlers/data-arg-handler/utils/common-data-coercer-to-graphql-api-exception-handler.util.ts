import { assertUnreachable } from 'twenty-shared/utils';

import {
  type CommonDataCoercerException,
  CommonDataCoercerExceptionCode,
} from 'src/engine/api/common/common-args-handlers/data-arg-handler/errors/common-data-coercer.exception';
import { UserInputError } from 'src/engine/core-modules/graphql/utils/graphql-errors.util';

export const commonDataCoercerToGraphqlApiExceptionHandler = (
  error: CommonDataCoercerException,
) => {
  switch (error.code) {
    case CommonDataCoercerExceptionCode.INVALID_COMPOSITE_FIELD:
    case CommonDataCoercerExceptionCode.INVALID_NUMBER:
    case CommonDataCoercerExceptionCode.INVALID_TEXT:
    case CommonDataCoercerExceptionCode.INVALID_DATE_OR_DATE_TIME:
    case CommonDataCoercerExceptionCode.INVALID_BOOLEAN:
    case CommonDataCoercerExceptionCode.INVALID_SELECT:
    case CommonDataCoercerExceptionCode.INVALID_MULTI_SELECT:
    case CommonDataCoercerExceptionCode.INVALID_UUID:
    case CommonDataCoercerExceptionCode.INVALID_ARRAY:
    case CommonDataCoercerExceptionCode.INVALID_RAW_JSON:
      throw new UserInputError(error);
    default: {
      return assertUnreachable(error.code);
    }
  }
};
