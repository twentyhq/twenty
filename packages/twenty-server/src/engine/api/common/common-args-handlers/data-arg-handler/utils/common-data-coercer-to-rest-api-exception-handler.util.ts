import { BadRequestException } from '@nestjs/common';

import { assertUnreachable } from 'twenty-shared/utils';

import {
  type CommonDataCoercerException,
  CommonDataCoercerExceptionCode,
} from 'src/engine/api/common/common-args-handlers/data-arg-handler/errors/common-data-coercer.exception';

export const commonDataCoercerToRestApiExceptionHandler = (
  error: CommonDataCoercerException,
) => {
  switch (error.code) {
    case CommonDataCoercerExceptionCode.INVALID_COMPOSITE_FIELD:
    case CommonDataCoercerExceptionCode.INVALID_NUMBER:
    case CommonDataCoercerExceptionCode.INVALID_TEXT:
    case CommonDataCoercerExceptionCode.INVALID_DATE_OR_DATE_TIME:
    case CommonDataCoercerExceptionCode.INVALID_BOOLEAN:
    case CommonDataCoercerExceptionCode.INVALID_SELECT:
    case CommonDataCoercerExceptionCode.INVALID_MULTI_SELECT:
    case CommonDataCoercerExceptionCode.INVALID_UUID:
    case CommonDataCoercerExceptionCode.INVALID_ARRAY:
    case CommonDataCoercerExceptionCode.INVALID_RAW_JSON:
      throw new BadRequestException(error.message);
    default: {
      return assertUnreachable(error.code);
    }
  }
};
