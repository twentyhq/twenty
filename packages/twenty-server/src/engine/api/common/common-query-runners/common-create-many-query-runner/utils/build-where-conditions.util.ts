import { isDefined } from 'twenty-shared/utils';
import { type FindOperator, In } from 'typeorm';

import { type ObjectRecord } from 'src/engine/api/graphql/workspace-query-builder/interfaces/object-record.interface';

import { getValueFromPath } from 'src/engine/api/common/common-query-runners/common-create-many-query-runner/utils/get-value-from-path.util';

export const buildWhereConditions = (
  records: Partial<ObjectRecord>[],
  conflictingFields: {
    baseField: string;
    fullPath: string;
    column: string;
  }[],
): Record<string, FindOperator<string>>[] => {
  const whereConditions = [];

  for (const field of conflictingFields) {
    const fieldValues = records
      .map((record) => getValueFromPath(record, field.fullPath))
      .filter(isDefined);

    if (fieldValues.length > 0) {
      whereConditions.push({ [field.column]: In(fieldValues) });
    }
  }

  return whereConditions;
};
