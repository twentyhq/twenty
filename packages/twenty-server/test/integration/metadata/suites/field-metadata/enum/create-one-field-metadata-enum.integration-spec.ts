import { createOneFieldMetadata } from 'test/integration/metadata/suites/field-metadata/utils/create-one-field-metadata.util';
import {
  LISTING_NAME_PLURAL,
  LISTING_NAME_SINGULAR,
} from 'test/integration/metadata/suites/object-metadata/constants/test-object-names.constant';
import { deleteOneObjectMetadata } from 'test/integration/metadata/suites/object-metadata/utils/delete-one-object-metadata.util';
import { forceCreateOneObjectMetadata } from 'test/integration/metadata/suites/object-metadata/utils/force-create-one-object-metadata.util';
import { CREATE_ENUM_FIELD_METADATA_TEST_CASES } from 'test/integration/metadata/suites/field-metadata/enum/create-enum-field-metadata-test-cases';
import { isDefined } from 'twenty-shared/utils';

import {
  FieldMetadataComplexOption,
  FieldMetadataDefaultOption,
} from 'src/engine/metadata-modules/field-metadata/dtos/options.input';
import { fieldMetadataEnumTypes } from 'src/engine/metadata-modules/field-metadata/utils/is-enum-field-metadata-type.util';

describe.each(fieldMetadataEnumTypes)(
  'Create field metadata %s tests suite',
  (testedFieldMetadataType) => {
    let createdObjectMetadataId: string;
    const testCases =
      CREATE_ENUM_FIELD_METADATA_TEST_CASES[testedFieldMetadataType];

    if (!isDefined(testCases)) {
      return;
    }
    const { failing: failingTestCases, successful: successfulTestCases } =
      testCases;

    beforeEach(async () => {
      const { data } = await forceCreateOneObjectMetadata({
        input: {
          labelSingular: LISTING_NAME_SINGULAR,
          labelPlural: LISTING_NAME_PLURAL,
          nameSingular: LISTING_NAME_SINGULAR,
          namePlural: LISTING_NAME_PLURAL,
          icon: 'IconBuildingSkyscraper',
          isLabelSyncedWithName: false,
        },
      });

      createdObjectMetadataId = data.createOneObject.id;
    });

    afterEach(async () => {
      await deleteOneObjectMetadata({
        input: { idToDelete: createdObjectMetadataId },
      });
    });

    test.each(successfulTestCases)(
      'Create $title',
      async ({ context: { input, expectedOptions } }) => {
        const { data, errors } = await createOneFieldMetadata({
          input: {
            objectMetadataId: createdObjectMetadataId,
            type: testedFieldMetadataType,
            name: 'testField',
            label: 'Test Field',
            isLabelSyncedWithName: false,
            ...input,
          },
          gqlFields: `
        id
        options
        defaultValue
        type
        `,
        });

        expect(data).not.toBeNull();
        expect(data.createOneField).toBeDefined();
        expect(data.createOneField.type).toEqual(testedFieldMetadataType);
        const createdOptions:
          | FieldMetadataDefaultOption[]
          | FieldMetadataComplexOption[] = data.createOneField.options;

        const optionsToCompare = expectedOptions ?? input.options;

        expect(errors).toBeUndefined();
        expect(createdOptions.length).toBe(optionsToCompare.length);
        createdOptions.forEach((option) => expect(option.id).toBeDefined());
        expect(createdOptions).toMatchObject(optionsToCompare);

        if (isDefined(input.defaultValue)) {
          expect(data.createOneField.defaultValue).toEqual(input.defaultValue);
        }
      },
    );

    test.each(failingTestCases)(
      'Create $title',
      async ({ context: { input } }) => {
        const { data, errors } = await createOneFieldMetadata({
          input: {
            objectMetadataId: createdObjectMetadataId,
            type: testedFieldMetadataType,
            name: 'testField',
            label: 'Test Field',
            isLabelSyncedWithName: false,
            ...input,
          },
          gqlFields: `
            id
            options
            defaultValue
            `,
        });

        expect(data).toBeNull();
        expect(errors).toBeDefined();
        expect(errors).toMatchSnapshot();
      },
    );
  },
);
