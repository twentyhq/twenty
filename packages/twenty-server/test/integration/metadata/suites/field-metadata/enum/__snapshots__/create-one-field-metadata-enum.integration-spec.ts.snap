// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create field metadata MULTI_SELECT tests suite Create should fail with an invalid default value 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "subCode": "INVALID_FIELD_INPUT",
      "userFriendlyMessage": "An error occurred.",
    },
    "message": "Default value for multi-select must be an array",
    "name": "UserInputError",
  },
]
`;

exports[`Create field metadata MULTI_SELECT tests suite Create should fail with an unknown default value 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "subCode": "INVALID_FIELD_INPUT",
      "userFriendlyMessage": "An error occurred.",
    },
    "message": "Default value for multi-select must be an array",
    "name": "UserInputError",
  },
]
`;

exports[`Create field metadata MULTI_SELECT tests suite Create should fail with comma in option label 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "subCode": "INVALID_FIELD_INPUT",
      "userFriendlyMessage": "Label must not contain a comma",
    },
    "message": "Label must not contain a comma",
    "name": "UserInputError",
  },
]
`;

exports[`Create field metadata MULTI_SELECT tests suite Create should fail with duplicated option ids 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "subCode": "INVALID_FIELD_INPUT",
      "userFriendlyMessage": "Duplicated option id",
    },
    "message": "Duplicated option id",
    "name": "UserInputError",
  },
]
`;

exports[`Create field metadata MULTI_SELECT tests suite Create should fail with duplicated option positions 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "subCode": "INVALID_FIELD_INPUT",
      "userFriendlyMessage": "Duplicated option position",
    },
    "message": "Duplicated option position",
    "name": "UserInputError",
  },
]
`;

exports[`Create field metadata MULTI_SELECT tests suite Create should fail with duplicated option values 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "subCode": "INVALID_FIELD_INPUT",
      "userFriendlyMessage": "Duplicated option value",
    },
    "message": "Duplicated option value",
    "name": "UserInputError",
  },
]
`;

exports[`Create field metadata MULTI_SELECT tests suite Create should fail with duplicated trimmed option values 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "subCode": "INVALID_FIELD_INPUT",
      "userFriendlyMessage": "Duplicated option value",
    },
    "message": "Duplicated option value",
    "name": "UserInputError",
  },
]
`;

exports[`Create field metadata MULTI_SELECT tests suite Create should fail with empty multi-select default values array 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "subCode": "INVALID_FIELD_INPUT",
      "userFriendlyMessage": "If defined default value must contain at least one value",
    },
    "message": "If defined default value must contain at least one value",
    "name": "UserInputError",
  },
]
`;

exports[`Create field metadata MULTI_SELECT tests suite Create should fail with empty options 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "subCode": "INVALID_FIELD_INPUT",
      "userFriendlyMessage": "An error occurred.",
    },
    "message": "Options are required for enum fields",
    "name": "UserInputError",
  },
]
`;

exports[`Create field metadata MULTI_SELECT tests suite Create should fail with empty string defaultValue 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "subCode": "INVALID_FIELD_INPUT",
      "userFriendlyMessage": "An error occurred.",
    },
    "message": "Default value for multi-select must be an array",
    "name": "UserInputError",
  },
]
`;

exports[`Create field metadata MULTI_SELECT tests suite Create should fail with empty string id 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "subCode": "INVALID_FIELD_INPUT",
      "userFriendlyMessage": "Option id is invalid",
    },
    "message": "Option id is invalid",
    "name": "UserInputError",
  },
]
`;

exports[`Create field metadata MULTI_SELECT tests suite Create should fail with empty string label 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "subCode": "INVALID_FIELD_INPUT",
      "userFriendlyMessage": "Option label "" is beneath 1 character",
    },
    "message": "Option label "" is beneath 1 character",
    "name": "UserInputError",
  },
]
`;

exports[`Create field metadata MULTI_SELECT tests suite Create should fail with empty string value 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "subCode": "INVALID_FIELD_INPUT",
      "userFriendlyMessage": "Option value "" is beneath 1 character",
    },
    "message": "Option value "" is beneath 1 character",
    "name": "UserInputError",
  },
]
`;

exports[`Create field metadata MULTI_SELECT tests suite Create should fail with invalid multi-select default value format 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "subCode": "INVALID_FIELD_INPUT",
      "userFriendlyMessage": "Default value should be as quoted string",
    },
    "message": "Default value should be as quoted string",
    "name": "UserInputError",
  },
]
`;

exports[`Create field metadata MULTI_SELECT tests suite Create should fail with invalid option id 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "subCode": "INVALID_FIELD_INPUT",
      "userFriendlyMessage": "Option id is invalid",
    },
    "message": "Option id is invalid",
    "name": "UserInputError",
  },
]
`;

exports[`Create field metadata MULTI_SELECT tests suite Create should fail with invalid option value format 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "subCode": "INVALID_FIELD_INPUT",
      "userFriendlyMessage": "Value must be in UPPER_CASE and follow snake_case "Option 1 and some other things, /"",
    },
    "message": "Value must be in UPPER_CASE and follow snake_case "Option 1 and some other things, /"",
    "name": "UserInputError",
  },
]
`;

exports[`Create field metadata MULTI_SELECT tests suite Create should fail with non stringified array default value 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "subCode": "INVALID_FIELD_INPUT",
      "userFriendlyMessage": "An error occurred.",
    },
    "message": "Default value for multi-select must be an array",
    "name": "UserInputError",
  },
]
`;

exports[`Create field metadata MULTI_SELECT tests suite Create should fail with not a string defaultValue 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "subCode": "INVALID_FIELD_INPUT",
      "userFriendlyMessage": "An error occurred.",
    },
    "message": "Default value for multi-select must be an array",
    "name": "UserInputError",
  },
]
`;

exports[`Create field metadata MULTI_SELECT tests suite Create should fail with not a string id 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "subCode": "INVALID_FIELD_INPUT",
      "userFriendlyMessage": "Option id is invalid",
    },
    "message": "Option id is invalid",
    "name": "UserInputError",
  },
]
`;

exports[`Create field metadata MULTI_SELECT tests suite Create should fail with not a string label 1`] = `
[
  {
    "extensions": {
      "code": "INTERNAL_SERVER_ERROR",
      "exceptionEventId": "mocked-exception-id",
      "userFriendlyMessage": "An error occurred.",
    },
    "message": "label.includes is not a function",
  },
]
`;

exports[`Create field metadata MULTI_SELECT tests suite Create should fail with not a string value 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "subCode": "INVALID_FIELD_INPUT",
      "userFriendlyMessage": "Value must be in UPPER_CASE and follow snake_case "22222"",
    },
    "message": "Value must be in UPPER_CASE and follow snake_case "22222"",
    "name": "UserInputError",
  },
]
`;

exports[`Create field metadata MULTI_SELECT tests suite Create should fail with null id 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "subCode": "INVALID_FIELD_INPUT",
      "userFriendlyMessage": "Option id is required",
    },
    "message": "Option id is required",
    "name": "UserInputError",
  },
]
`;

exports[`Create field metadata MULTI_SELECT tests suite Create should fail with null label 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "subCode": "INVALID_FIELD_INPUT",
      "userFriendlyMessage": "Option label is required",
    },
    "message": "Option label is required",
    "name": "UserInputError",
  },
]
`;

exports[`Create field metadata MULTI_SELECT tests suite Create should fail with null options 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "subCode": "INVALID_FIELD_INPUT",
      "userFriendlyMessage": "An error occurred.",
    },
    "message": "Options are required for enum fields",
    "name": "UserInputError",
  },
]
`;

exports[`Create field metadata MULTI_SELECT tests suite Create should fail with null value 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "subCode": "INVALID_FIELD_INPUT",
      "userFriendlyMessage": "Option value is required",
    },
    "message": "Option value is required",
    "name": "UserInputError",
  },
]
`;

exports[`Create field metadata MULTI_SELECT tests suite Create should fail with only white spaces defaultValue 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "subCode": "INVALID_FIELD_INPUT",
      "userFriendlyMessage": "An error occurred.",
    },
    "message": "Default value for multi-select must be an array",
    "name": "UserInputError",
  },
]
`;

exports[`Create field metadata MULTI_SELECT tests suite Create should fail with only white spaces id 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "subCode": "INVALID_FIELD_INPUT",
      "userFriendlyMessage": "Option id is invalid",
    },
    "message": "Option id is invalid",
    "name": "UserInputError",
  },
]
`;

exports[`Create field metadata MULTI_SELECT tests suite Create should fail with only white spaces label 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "subCode": "INVALID_FIELD_INPUT",
      "userFriendlyMessage": "Option label "" is beneath 1 character",
    },
    "message": "Option label "" is beneath 1 character",
    "name": "UserInputError",
  },
]
`;

exports[`Create field metadata MULTI_SELECT tests suite Create should fail with only white spaces value 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "subCode": "INVALID_FIELD_INPUT",
      "userFriendlyMessage": "Option value "" is beneath 1 character",
    },
    "message": "Option value "" is beneath 1 character",
    "name": "UserInputError",
  },
]
`;

exports[`Create field metadata MULTI_SELECT tests suite Create should fail with too long defaultValue 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "subCode": "INVALID_FIELD_INPUT",
      "userFriendlyMessage": "An error occurred.",
    },
    "message": "Default value for multi-select must be an array",
    "name": "UserInputError",
  },
]
`;

exports[`Create field metadata MULTI_SELECT tests suite Create should fail with too long id 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "subCode": "INVALID_FIELD_INPUT",
      "userFriendlyMessage": "Option id is invalid",
    },
    "message": "Option id is invalid",
    "name": "UserInputError",
  },
]
`;

exports[`Create field metadata MULTI_SELECT tests suite Create should fail with too long label 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "subCode": "INVALID_FIELD_INPUT",
      "userFriendlyMessage": "Option label exceeds 63 characters",
    },
    "message": "Option label exceeds 63 characters",
    "name": "UserInputError",
  },
]
`;

exports[`Create field metadata MULTI_SELECT tests suite Create should fail with too long value 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "subCode": "INVALID_FIELD_INPUT",
      "userFriendlyMessage": "Option value exceeds 63 characters",
    },
    "message": "Option value exceeds 63 characters",
    "name": "UserInputError",
  },
]
`;

exports[`Create field metadata MULTI_SELECT tests suite Create should fail with undefined option label 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "subCode": "INVALID_FIELD_INPUT",
      "userFriendlyMessage": "Option label is required",
    },
    "message": "Option label is required",
    "name": "UserInputError",
  },
]
`;

exports[`Create field metadata MULTI_SELECT tests suite Create should fail with undefined option value 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "subCode": "INVALID_FIELD_INPUT",
      "userFriendlyMessage": "Option value is required",
    },
    "message": "Option value is required",
    "name": "UserInputError",
  },
]
`;

exports[`Create field metadata MULTI_SELECT tests suite Create should fail with undefined options 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "subCode": "INVALID_FIELD_INPUT",
      "userFriendlyMessage": "An error occurred.",
    },
    "message": "Options are required for enum fields",
    "name": "UserInputError",
  },
]
`;

exports[`Create field metadata MULTI_SELECT tests suite Create should fail with unknown multi-select default values 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "subCode": "INVALID_FIELD_INPUT",
      "userFriendlyMessage": "Default value "'UNKNOWN_OPTION'" must be one of the option values",
    },
    "message": "Default value "'UNKNOWN_OPTION'" must be one of the option values",
    "name": "UserInputError",
  },
]
`;

exports[`Create field metadata SELECT tests suite Create should fail with an invalid default value 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "subCode": "INVALID_FIELD_INPUT",
      "userFriendlyMessage": "Default value should be as quoted string",
    },
    "message": "Default value should be as quoted string",
    "name": "UserInputError",
  },
]
`;

exports[`Create field metadata SELECT tests suite Create should fail with an unknown default value 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "subCode": "INVALID_FIELD_INPUT",
      "userFriendlyMessage": "Default value "'OPTION_424242'" must be one of the option values",
    },
    "message": "Default value "'OPTION_424242'" must be one of the option values",
    "name": "UserInputError",
  },
]
`;

exports[`Create field metadata SELECT tests suite Create should fail with comma in option label 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "subCode": "INVALID_FIELD_INPUT",
      "userFriendlyMessage": "Label must not contain a comma",
    },
    "message": "Label must not contain a comma",
    "name": "UserInputError",
  },
]
`;

exports[`Create field metadata SELECT tests suite Create should fail with duplicated option ids 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "subCode": "INVALID_FIELD_INPUT",
      "userFriendlyMessage": "Duplicated option id",
    },
    "message": "Duplicated option id",
    "name": "UserInputError",
  },
]
`;

exports[`Create field metadata SELECT tests suite Create should fail with duplicated option positions 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "subCode": "INVALID_FIELD_INPUT",
      "userFriendlyMessage": "Duplicated option position",
    },
    "message": "Duplicated option position",
    "name": "UserInputError",
  },
]
`;

exports[`Create field metadata SELECT tests suite Create should fail with duplicated option values 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "subCode": "INVALID_FIELD_INPUT",
      "userFriendlyMessage": "Duplicated option value",
    },
    "message": "Duplicated option value",
    "name": "UserInputError",
  },
]
`;

exports[`Create field metadata SELECT tests suite Create should fail with duplicated trimmed option values 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "subCode": "INVALID_FIELD_INPUT",
      "userFriendlyMessage": "Duplicated option value",
    },
    "message": "Duplicated option value",
    "name": "UserInputError",
  },
]
`;

exports[`Create field metadata SELECT tests suite Create should fail with empty options 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "subCode": "INVALID_FIELD_INPUT",
      "userFriendlyMessage": "An error occurred.",
    },
    "message": "Options are required for enum fields",
    "name": "UserInputError",
  },
]
`;

exports[`Create field metadata SELECT tests suite Create should fail with empty string defaultValue 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "subCode": "INVALID_FIELD_INPUT",
      "userFriendlyMessage": "Default value should be as quoted string",
    },
    "message": "Default value should be as quoted string",
    "name": "UserInputError",
  },
]
`;

exports[`Create field metadata SELECT tests suite Create should fail with empty string id 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "subCode": "INVALID_FIELD_INPUT",
      "userFriendlyMessage": "Option id is invalid",
    },
    "message": "Option id is invalid",
    "name": "UserInputError",
  },
]
`;

exports[`Create field metadata SELECT tests suite Create should fail with empty string label 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "subCode": "INVALID_FIELD_INPUT",
      "userFriendlyMessage": "Option label "" is beneath 1 character",
    },
    "message": "Option label "" is beneath 1 character",
    "name": "UserInputError",
  },
]
`;

exports[`Create field metadata SELECT tests suite Create should fail with empty string value 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "subCode": "INVALID_FIELD_INPUT",
      "userFriendlyMessage": "Option value "" is beneath 1 character",
    },
    "message": "Option value "" is beneath 1 character",
    "name": "UserInputError",
  },
]
`;

exports[`Create field metadata SELECT tests suite Create should fail with invalid option id 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "subCode": "INVALID_FIELD_INPUT",
      "userFriendlyMessage": "Option id is invalid",
    },
    "message": "Option id is invalid",
    "name": "UserInputError",
  },
]
`;

exports[`Create field metadata SELECT tests suite Create should fail with invalid option value format 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "subCode": "INVALID_FIELD_INPUT",
      "userFriendlyMessage": "Value must be in UPPER_CASE and follow snake_case "Option 1 and some other things, /"",
    },
    "message": "Value must be in UPPER_CASE and follow snake_case "Option 1 and some other things, /"",
    "name": "UserInputError",
  },
]
`;

exports[`Create field metadata SELECT tests suite Create should fail with not a string defaultValue 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "subCode": "INVALID_FIELD_INPUT",
      "userFriendlyMessage": "An error occurred.",
    },
    "message": "Default value for multi-select must be a stringified array",
    "name": "UserInputError",
  },
]
`;

exports[`Create field metadata SELECT tests suite Create should fail with not a string id 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "subCode": "INVALID_FIELD_INPUT",
      "userFriendlyMessage": "Option id is invalid",
    },
    "message": "Option id is invalid",
    "name": "UserInputError",
  },
]
`;

exports[`Create field metadata SELECT tests suite Create should fail with not a string label 1`] = `
[
  {
    "extensions": {
      "code": "INTERNAL_SERVER_ERROR",
      "exceptionEventId": "mocked-exception-id",
      "userFriendlyMessage": "An error occurred.",
    },
    "message": "label.includes is not a function",
  },
]
`;

exports[`Create field metadata SELECT tests suite Create should fail with not a string value 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "subCode": "INVALID_FIELD_INPUT",
      "userFriendlyMessage": "Value must be in UPPER_CASE and follow snake_case "22222"",
    },
    "message": "Value must be in UPPER_CASE and follow snake_case "22222"",
    "name": "UserInputError",
  },
]
`;

exports[`Create field metadata SELECT tests suite Create should fail with null id 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "subCode": "INVALID_FIELD_INPUT",
      "userFriendlyMessage": "Option id is required",
    },
    "message": "Option id is required",
    "name": "UserInputError",
  },
]
`;

exports[`Create field metadata SELECT tests suite Create should fail with null label 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "subCode": "INVALID_FIELD_INPUT",
      "userFriendlyMessage": "Option label is required",
    },
    "message": "Option label is required",
    "name": "UserInputError",
  },
]
`;

exports[`Create field metadata SELECT tests suite Create should fail with null options 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "subCode": "INVALID_FIELD_INPUT",
      "userFriendlyMessage": "An error occurred.",
    },
    "message": "Options are required for enum fields",
    "name": "UserInputError",
  },
]
`;

exports[`Create field metadata SELECT tests suite Create should fail with null value 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "subCode": "INVALID_FIELD_INPUT",
      "userFriendlyMessage": "Option value is required",
    },
    "message": "Option value is required",
    "name": "UserInputError",
  },
]
`;

exports[`Create field metadata SELECT tests suite Create should fail with only white spaces defaultValue 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "subCode": "INVALID_FIELD_INPUT",
      "userFriendlyMessage": "Default value should be as quoted string",
    },
    "message": "Default value should be as quoted string",
    "name": "UserInputError",
  },
]
`;

exports[`Create field metadata SELECT tests suite Create should fail with only white spaces id 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "subCode": "INVALID_FIELD_INPUT",
      "userFriendlyMessage": "Option id is invalid",
    },
    "message": "Option id is invalid",
    "name": "UserInputError",
  },
]
`;

exports[`Create field metadata SELECT tests suite Create should fail with only white spaces label 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "subCode": "INVALID_FIELD_INPUT",
      "userFriendlyMessage": "Option label "" is beneath 1 character",
    },
    "message": "Option label "" is beneath 1 character",
    "name": "UserInputError",
  },
]
`;

exports[`Create field metadata SELECT tests suite Create should fail with only white spaces value 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "subCode": "INVALID_FIELD_INPUT",
      "userFriendlyMessage": "Option value "" is beneath 1 character",
    },
    "message": "Option value "" is beneath 1 character",
    "name": "UserInputError",
  },
]
`;

exports[`Create field metadata SELECT tests suite Create should fail with too long defaultValue 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "subCode": "INVALID_FIELD_INPUT",
      "userFriendlyMessage": "Default value should be as quoted string",
    },
    "message": "Default value should be as quoted string",
    "name": "UserInputError",
  },
]
`;

exports[`Create field metadata SELECT tests suite Create should fail with too long id 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "subCode": "INVALID_FIELD_INPUT",
      "userFriendlyMessage": "Option id is invalid",
    },
    "message": "Option id is invalid",
    "name": "UserInputError",
  },
]
`;

exports[`Create field metadata SELECT tests suite Create should fail with too long label 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "subCode": "INVALID_FIELD_INPUT",
      "userFriendlyMessage": "Option label exceeds 63 characters",
    },
    "message": "Option label exceeds 63 characters",
    "name": "UserInputError",
  },
]
`;

exports[`Create field metadata SELECT tests suite Create should fail with too long value 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "subCode": "INVALID_FIELD_INPUT",
      "userFriendlyMessage": "Option value exceeds 63 characters",
    },
    "message": "Option value exceeds 63 characters",
    "name": "UserInputError",
  },
]
`;

exports[`Create field metadata SELECT tests suite Create should fail with undefined option label 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "subCode": "INVALID_FIELD_INPUT",
      "userFriendlyMessage": "Option label is required",
    },
    "message": "Option label is required",
    "name": "UserInputError",
  },
]
`;

exports[`Create field metadata SELECT tests suite Create should fail with undefined option value 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "subCode": "INVALID_FIELD_INPUT",
      "userFriendlyMessage": "Option value is required",
    },
    "message": "Option value is required",
    "name": "UserInputError",
  },
]
`;

exports[`Create field metadata SELECT tests suite Create should fail with undefined options 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "subCode": "INVALID_FIELD_INPUT",
      "userFriendlyMessage": "An error occurred.",
    },
    "message": "Options are required for enum fields",
    "name": "UserInputError",
  },
]
`;